### DISCUSSION TITLE
What is the best code architecture for huge forms?

### DISCUSSION THREAD

**Discussion by AkselAllas** [react-hook-form/react-hook-form#3055] [Q&A]
I have a similar question to:
https://stackoverflow.com/questions/54850281/react-dividing-a-large-form-into-multiple-components

When I have ~40 inputs and I would get to ~1000 lines of code, then what is the best way to structure my code? 

I'm thinking about two choices:

1) I make Controlled Child Components using Controller and having that Child Component pass up only valid values in the style of `{input1: input1Value, input2: input2Value, input3: input3Value}`. The downsides to this is the double rendering of those inputs and having to spread the new object passed up in an `onSubmit`. 

2) I make everything flat in one file, which goes 1000+ lines.

**Comment by AkselAllas** [comment]
I'm gonna try using FormProvider

**Comment by HasanMothaffar** [comment]
I went with the second option and ended up with 1500+ lines.

Most of my controllers are between 20-25 lines as in the image below (I'm using MUI). I thought of creating abstractions but figured they would be a pain to maintain / hard to understand in the future.

![image](https://github.com/react-hook-form/react-hook-form/assets/60794869/206788a0-c28c-4466-8900-8cbaea65c7f9)

@AkselAllas How was your experience with using FormProvider?

**Reply by Moshyfawn** [reply]
The abstraction is already written 👌
https://github.com/dohomi/react-hook-form-mui

**Reply by AkselAllas** [reply]
> How was your experience with using FormProvider?

I sadly don't remember. I'm no longer doing frontend 😂 🙃

**Reply by HasanMothaffar** [reply]
@Moshyfawn Thanks! It seems like an interesting project, I'll look into it. The only problem is that it's community-driven.. so I'm not sure if it stays up to date when either RHF or MUI release updates/breaking changes.

**Reply by HasanMothaffar** [reply]
@AkselAllas I think I'll follow along haha

**Reply by Moshyfawn** [reply]
> I'm not sure if it stays up to date when either RHF or MUI release updates/breaking changes.

They've been pretty good at following along.

**Reply by HasanMothaffar** [reply]
I found an interesting abstraction using TypeScript on this [reddit thread](https://www.reddit.com/r/reactjs/comments/14f3gcn/comment/jozc5uh/?utm_source=reddit&utm_medium=web2x&context=3) in case someone's interested.
